### PR TITLE
Further improving resolve vector notation

### PR DESCRIPTION
**Pull request by MichaelSt98**
#### Summary
This PR improves the robustness of vector-notation resolution by making it more conservative where transformations are ambiguous, while still preserving important scalarizable cases that appear in real physics code.

#### Key Changes

- Improves handling of inline function calls so elemental cases are recognized more reliably during vector-notation resolution.
- Refines how slice bounds are analyzed to distinguish safe scalar bounds from unsafe vector-valued expressions.
- Restores valid transformations for scalar derived-type bounds while keeping unsafe unresolved cases visible through warnings.
- Prevents incorrect resolution when array subscripts or bounds are effectively vector-valued.
- Expands regression coverage for representative IFS and ecRad-style patterns.
- Cleans up the related test setup so frontend-generated artifacts do not leak into the repository root.

#### Outcome

Overall, the diff makes resolve_vector_notation safer and more predictable: it avoids wrong-code generation in ambiguous cases, fixes missed transformations in valid cases, and adds coverage for the boundary cases that triggered the regressions.

### Contributor Declaration

By opening this pull request, I affirm the following:

* All authors agree to the [Contributor License Agreement](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md).
* The code follows the project's coding standards.
* I have performed self-review and added comments where needed.
* I have added or updated tests to verify that my changes are effective and functional.
* I have run all existing tests and confirmed they pass.

### CODE REVIEW

File: loki/transformations/array_indexing/tests/test_vector_notation.py [hunk 1]
```diff
@@ -1680,6 +1838,437 @@ def test_resolve_vector_notation_shifted_range_multi_arg(frontend):
...
+    integer :: band(3)
+  end type config_type
+contains
+  subroutine test_vector_slice_bounds(config, arr, out)
+    implicit none
+    type(config_type), intent(in) :: config
+    real, intent(in) :: arr(5)
+    real, intent(out) :: out(5)
+
+    out(config%band:5) = arr(config%band:5)
```
Comment (awnawab) [inline loki/transformations/array_indexing/tests/test_vector_notation.py]: Is this even valid Fortran?

File: loki/transformations/array_indexing/tests/test_vector_notation.py [hunk 2]
```diff
@@ -1680,6 +1838,437 @@ def test_resolve_vector_notation_shifted_range_multi_arg(frontend):
...
+    """
+    from loki.logging import WARNING  # pylint: disable=import-outside-toplevel
+
+    fcode = """
+subroutine test_unresolved_scalar_bound_warning(arr, out)
+  implicit none
+  real, intent(in) :: arr(5)
+  real, intent(out) :: out(5)
+
+  out(1:dims%klon) = arr(1:dims%klon)
```
Comment (awnawab) [inline loki/transformations/array_indexing/tests/test_vector_notation.py]: If `dims%klon` were anything but a scalar, this wouldn't be valid fortran, so if we are slicing a dimension, the assumption of a scalar is a safe one.

File: loki/transformations/array_indexing/vector_notation.py [hunk 3]
```diff
@@ -313,6 +313,18 @@ def _shape_to_range(s):
...
+    @staticmethod
+    def _shape_lower(s):
+        return s.lower if isinstance(s, sym.Range) else sym.IntLiteral(1)
+
+    @staticmethod
+    def _shape_upper(s):
+        return s.upper if isinstance(s, sym.Range) else s
+
+    @staticmethod
+    def _shape_step(s):
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: A dimension can have a step, but can a shape even have a step? for example, I think this declaration would be invalid:
```
real :: a(1:5:2)
```

File: loki/transformations/array_indexing/vector_notation.py [hunk 4]
```diff
@@ -419,10 +441,320 @@ def _find_qualified_range_positions(dims, range_positions):
         Return ordinal indices into ``range_positions`` whose corresponding
         dimension is *not* a bare ``(:)`` (i.e. ``RangeIndex((None, None))``).
         """
+        return ResolveVectorNotationTransformer._find_resolvable_range_positions(dims, range_positions)
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: Is there a point in making this static if we then have to access a method via the class name?

File: loki/transformations/array_indexing/vector_notation.py [hunk 5]
```diff
@@ -419,10 +441,320 @@ def _find_qualified_range_positions(dims, range_positions):
...
+
+    @staticmethod
+    def _has_explicit_range_bounds(dim):
+        return isinstance(dim, sym.RangeIndex) and dim.lower is not None and dim.upper is not None
+
+    @staticmethod
+    def _warn_on_unresolved_scalarizable_bound(bound):
+        if any(isinstance(var, sym.DeferredTypeSymbol) and var.parent is not None
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: If the bound itself is a deferred type symbol, isn't that all we need to know? So is the presence of changing the situation at all? consider:
```
a(c) = ...
a(b%c) = ...
```

Whether `c` is deferredtype or `b%c` is deferredtype, the risk here would be the same.

File: loki/transformations/array_indexing/vector_notation.py [hunk 6]
```diff
@@ -419,10 +441,320 @@ def _find_qualified_range_positions(dims, range_positions):
...
+        wrapped_body = body
+        for ivar, irange in index_range_map.items():
+            if isinstance(irange, sym.RangeIndex):
+                bounds = sym.LoopRange(irange.children)
+            else:
+                bounds = sym.LoopRange((sym.Literal(1), irange, sym.Literal(1)))
+            loop = ir.Loop(variable=ivar, body=as_tuple(wrapped_body), bounds=bounds)
+            wrapped_body = loop
+
+        if insert_comments and loop is not None:
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: A consistent style here would make it more readable:
`if insert_comments and loop`
or
`if insert_comments is not None and loop is not None`

File: loki/transformations/array_indexing/vector_notation.py [hunk 7]
```diff
@@ -419,10 +441,320 @@ def _find_qualified_range_positions(dims, range_positions):
...
+            if scope is not None and hasattr(scope, 'subroutines'):
+                return next(
+                    (routine for routine in scope.subroutines
+                     if routine.name.lower() == call.name.lower()),
+                    None
+                )
+            return None
+
+        procedure_dtype = getattr(call.function.type, 'dtype', None)
+        if procedure_dtype is None or procedure_dtype is BasicType.DEFERRED:
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: `BasicType.DEFERRED` was recently made falsey to facilitate these kinds of checks, so this could just be:
`if not procedure_dtype`.

**Review by awnawab — CHANGES_REQUESTED** [review]
Thanks a lot for this great contribution @MichaelSt98 🙏 This makes the vector notation resolution much more powerful and will require fewer interventions in the source itself 👌 In fact, I'll hold my ifs-source PR until this is merged, as possibly now loki can handle some of  the cases I resolved manually.

Overall the changes look very good. I've just a few comments where the implementation can perhaps be tidied up a little, or where the warnings or tests can expose the underlying problem a little more clearly.

File: loki/transformations/array_indexing/vector_notation.py [hunk 8]
```diff
@@ -693,12 +1024,21 @@ def visit_Assignment(self, stmt, **kwargs):  # pylint: disable=unused-argument
...
-        # --- Step 5: Map LHS ranges to loop index variables ---
+        # --- Step 6: Map LHS ranges to loop index variables ---
         resolved_lhs_ranges = [
             lhs_dims[lhs_range_positions[i]] for i in resolvable_dim_indices
         ]
+        # Collect loop variables already present as fixed scalar subscripts so
+        # that _map_ranges_to_indices does not alias them with a resolved range.
+        reserved_ivars = {
+            d for i, d in enumerate(lhs_dims)
+            if i not in lhs_range_positions and isinstance(d, sym.Scalar) and d in self.loop_map.values()
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py:1111]: Since we haven't resolved any vector notation yet in this assignment, isn't `d in self.loop_map.values()` in itself sufficient? by definition, none of the loop vars will be in lhs_range_positions. It can't hurt to leave that in, but it makes that check more readable if we only check the contents of the loop_map imo.

File: loki/transformations/array_indexing/vector_notation.py [hunk 9]
```diff
@@ -693,12 +1024,21 @@ def visit_Assignment(self, stmt, **kwargs):  # pylint: disable=unused-argument
...
+        }
+        # Keep existing scalar subscripts and active surrounding loop indices
+        # distinct from any new loop variable chosen for resolved ranges.
+        reserved_ivars.update(self.active_loop_vars)
         new_lhs_dims, index_range_map, synthesized_ivars = self._map_ranges_to_indices(
-            resolved_lhs_ranges, self.loop_map,
+            resolved_lhs_ranges, self.loop_map, reserved_ivars=reserved_ivars,
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py:1117]: `map_unknown_ranges` is a little misleading, since we only resolve fully qualified ranges anyway. A better name might be `create_new_loop_indices`?

File: loki/transformations/array_indexing/vector_notation.py [hunk 10]
```diff
@@ -413,16 +432,325 @@ def _find_range_positions(dims):
...
+        if not resolvable_dim_indices:
+            return expr, {}, False
+
+        resolved_ranges = [
+            cond_dims_per_array[0][lhs_range_positions[i]] for i in resolvable_dim_indices
+        ]
+        cond_all_dims = cond_arrays[0].dimensions
+        reserved_ivars = {
+            d for i, d in enumerate(cond_all_dims)
+            if i not in lhs_range_positions and isinstance(d, sym.Scalar) and d in self.loop_map.values()
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py:729]: Same comment here about loop indices, `i not in lhs_range_positions` is possibly redundant.

File: loki/transformations/array_indexing/vector_notation.py [hunk 11]
```diff
@@ -831,59 +1166,28 @@ def visit_Conditional(self, o, **kwargs):
...
+            return masked
 
-        conditions = SubstituteExpressions(subst_map).visit(conditions)
+        condition, index_range_map, resolved = self._resolve_mask_expr(masked.conditions[0])
+        if not resolved:
+            return masked
 
         with dict_override(kwargs, {'create_loops': False}):
-            bodies = self.visit(masked.bodies, **kwargs)
+            body = self.visit(masked.bodies[0], **kwargs)
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: Because the expression substitution is different for the condition and body, we will end up using different loop variables unless the dims to be resolved already happen to be in self.loop_map (in most cases these would be the horiztonal dim for example). For example:
```
where (a(1:n) ..)
 b(1:n)
end where
```
if we don't have any loops over n in that subrouine and it's not the dim over which we call `resolve_vector_dimension`, then n won't be in self.loop_map, and the transformation will substitute a different loop index in the mask and body assignment. We could fix this by adding the `MaskedStatement`s index_range_map to self.loop_map before visiting the bodies.

File: loki/transformations/array_indexing/vector_notation.py [hunk 12]
```diff
@@ -413,16 +432,325 @@ def _find_range_positions(dims):
...
+            if routine is not None:
+                return routine.procedure_type.is_elemental
+            return False
+        if hasattr(procedure_dtype, 'is_elemental'):
+            if getattr(procedure_dtype, 'procedure', BasicType.DEFERRED) is BasicType.DEFERRED:
+                routine = resolve_routine_from_scope()
+                if routine is not None:
+                    return routine.procedure_type.is_elemental
+            return procedure_dtype.is_elemental
+        procedure_type = call.procedure_type
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: This last check is redundant. call.procedure_type is the same as `call.function.type.dtype`, and we've already checked whether that's deffered or not and whether it's elemental.

File: loki/transformations/array_indexing/vector_notation.py [hunk 13]
```diff
@@ -413,16 +432,325 @@ def _find_range_positions(dims):
...
+                return next(
+                    (routine for routine in scope.subroutines
+                     if routine.name.lower() == call.name.lower()),
+                    None
+                )
+            return None
+
+        procedure_dtype = getattr(call.function.type, 'dtype', None)
+        if not procedure_dtype:
+            routine = resolve_routine_from_scope()
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py:599]: We'll end up here for an unenriched inline call, but this is currently uncovered, so it would be good to add a test for that.

File: loki/transformations/array_indexing/vector_notation.py [hunk 14]
```diff
@@ -413,16 +432,325 @@ def _find_range_positions(dims):
...
+                if cls._expr_contains_vector_array(dim):
+                    return True
+        return False
+
+    def _has_unsafe_inline_call(self, expr):
+        """Return True if expr contains a non-scalarizable inline call with vector args."""
+        for call in FindInlineCalls().visit(expr):
+            if self._is_scalarizable_inline_call(call):
+                continue
+            if any(isinstance(arg, sym.Array) for arg in FindVariables(unique=False).visit(call)):
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py:675]: That's an edge case we might actually hit, an inline call with an array arg. Could you please add a test for this?

File: loki/transformations/array_indexing/vector_notation.py [hunk 15]
```diff
@@ -413,16 +432,325 @@ def _find_range_positions(dims):
...
+    @staticmethod
+    def _build_loop_nest(index_range_map, body, insert_comments=False):
+        """Wrap ``body`` in loops for all ranges from ``index_range_map``."""
+        loop = None
+        wrapped_body = body
+        for ivar, irange in index_range_map.items():
+            if isinstance(irange, sym.RangeIndex):
+                bounds = sym.LoopRange(irange.children)
+            else:
+                bounds = sym.LoopRange((sym.Literal(1), irange, sym.Literal(1)))
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: We always seem to construct a `RangeIndex` by the time we get here, so this branch can either be removed or replaced by an assert.

File: loki/transformations/array_indexing/vector_notation.py [hunk 16]
```diff
@@ -413,16 +432,325 @@ def _find_range_positions(dims):
...
+            loop = ir.Loop(variable=ivar, body=as_tuple(wrapped_body), bounds=bounds)
+            wrapped_body = loop
+
+        if insert_comments and loop:
+            return (ir.Comment('! loki resolved vector notation'), loop)
+        if loop:
+            return (loop,)
+        return body
```
Comment (awnawab) [inline loki/transformations/array_indexing/vector_notation.py]: I think we are guaranteed to always have `index_range_map` defined, so we'll always create a loop and this `return body` is redundant.

**Review by awnawab — CHANGES_REQUESTED** [review]
Many thanks for improving this already very powerful utility 🙏 I'm sorry for doing this review in a piecemeal fashion, but this deserved a closer look than I had given it first time round.

On the whole the changes are great, I just found one potential bug and a couple of minor implementaiton details that can be tidied up.

File: loki/transformations/array_indexing/vector_notation.py (hunk 8 above)
Comment (MichaelSt98) [reply loki/transformations/array_indexing/vector_notation.py:1111]: `lhs_range_positions` is needed to select only fixed scalar subscripts. Without it, a range dimension equal to a loop-map variable could incorrectly reserve that variable.

File: loki/transformations/array_indexing/vector_notation.py (hunk 10 above)
Comment (MichaelSt98) [reply loki/transformations/array_indexing/vector_notation.py:729]: See other comment.

File: loki/transformations/array_indexing/vector_notation.py (hunk 14 above)
Comment (MichaelSt98) [reply loki/transformations/array_indexing/vector_notation.py:675]: Should be covered by: `test_resolve_vector_notation_non_elemental_inline_call_is_conservative`

File: loki/transformations/array_indexing/vector_notation.py (hunk 8 above)
Comment (awnawab) [reply loki/transformations/array_indexing/vector_notation.py:1111]: Got it, thanks for the clarification.

**Review by awnawab — APPROVED** [review]
Thanks a lot @MichaelSt98 for addressing all the feedback 🙏 This is now G2G ✅

File: loki/transformations/array_indexing/tests/test_vector_notation.py [hunk 17]
```diff
@@ -804,6 +805,95 @@ def test_resolve_vector_notation_ifs_patterns(frontend):
...
+    first_assign = FindNodes(ir.Assignment).visit(loops[0].body)[0]
+    assert first_assign.lhs == 'ssa_total(jl)'
+    assert first_assign.rhs == '(ssa(jl) + cloud(jl))/od_total(jl)'
+
+    second_assign = FindNodes(ir.Assignment).visit(loops[1].body)[0]
+    assert second_assign.lhs == 'g_total(jl)'
+    assert second_assign.rhs == '(g(jl) + cloud(jl))/ssa_total(jl)'
+
+    second_inner_cond = next(c for c in FindNodes(ir.Conditional).visit(loops[1].body))
+    second_cond = str(second_inner_cond.condition).replace(' ', '').lower()
```
Comment (mlange05) [inline loki/transformations/array_indexing/tests/test_vector_notation.py]: String fuffing should not be needed here - does native expression comparison not work for comparison operators?

File: loki/transformations/array_indexing/tests/test_vector_notation.py [hunk 18]
```diff
@@ -1467,37 +1562,47 @@ def test_resolve_vector_notation_no_implicit_rhs(frontend):
...
+        "Only the assignment with bare RHS ':' should remain unresolved"
+    assert 'b(:)' in str(unresolved[0].rhs), \
+        f"Unexpected unresolved assignment: {unresolved[0]}"
+
+    # The resolved loops should cover explicit-range and scalar RHS assignments.
+    loop_assigns = [FindNodes(ir.Assignment).visit(loop.body)[0] for loop in loops]
+    rhs_texts = {str(assign.rhs).replace(' ', '') for assign in loop_assigns}
+    assert any('c(' in rhs for rhs in rhs_texts), \
+        f"Expected resolved c(...) assignment, got: {rhs_texts}"
+    assert '0.0' in rhs_texts, f"Expected scalar literal assignment, got: {rhs_texts}"
```
Comment (mlange05) [inline loki/transformations/array_indexing/tests/test_vector_notation.py:1595]: Direct symbol-comparison should definitely work here.

File: loki/transformations/array_indexing/tests/test_vector_notation.py [hunk 19]
```diff
@@ -1507,21 +1612,95 @@ def test_resolve_vector_notation_no_implicit_rhs(frontend):
...
+        f"Unexpected unresolved assignment: {unresolved2[0]}"
+
+    # The resolved loop bodies should reference explicit-range and scalar RHS cases.
+    loop_assigns2 = [FindNodes(ir.Assignment).visit(loop.body)[0] for loop in loops2]
+    rhs_texts2 = {str(assign.rhs).replace(' ', '') for assign in loop_assigns2}
+    assert any('c(' in rhs for rhs in rhs_texts2), \
+        f"Expected resolved c(...) assignment, got: {rhs_texts2}"
+    assert '0.0' in rhs_texts2, f"Expected scalar literal assignment, got: {rhs_texts2}"
+    assert 'scalar' in rhs_texts2, f"Expected scalar variable assignment, got: {rhs_texts2}"
+    assert all(str(assign.lhs) == 'a(1, jl)' for assign in loop_assigns2)
```
Comment (mlange05) [inline loki/transformations/array_indexing/tests/test_vector_notation.py]: Please prefer direct comparison over string-casting

File: loki/transformations/array_indexing/tests/test_vector_notation.py [hunk 20]
```diff
@@ -1507,21 +1612,95 @@ def test_resolve_vector_notation_no_implicit_rhs(frontend):
...
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_resolve_masked_statements_scalar_rhs_with_no_implicit_rhs(frontend):
```
Comment (mlange05) [inline loki/transformations/array_indexing/tests/test_vector_notation.py]: Isn't this the same test code as in `test_resolve_masked_statements_fallback_unresolved_rhs`? Could these two be merged and the differing flag-output pairs be parametrised?

File: loki/transformations/single_column/tests/test_scc.py [hunk 21]
```diff
@@ -204,6 +204,151 @@ def test_scc_base_resolve_vector_notation_config(frontend, horizontal):
...
+    assert 'jl' in kernel2.variables
+    assert not FindNodes(MaskedStatement).visit(kernel2.body)
+    loops = FindNodes(Loop).visit(kernel2.body)
+    assert len(loops) == 1
+    assert loops[0].variable == 'jl'
+    assert loops[0].bounds == 'start:end'
+
+    conds = FindNodes(Conditional).visit(kernel2.body)
+    assert len(conds) == 1
+    assert conds[0].condition == 'q(jl) > 0.'
```
Comment (mlange05) [inline loki/transformations/single_column/tests/test_scc.py:263]: Should we also test the correct `Assignment` adjustment here?

File: loki/transformations/single_column/tests/test_scc.py [hunk 22]
```diff
@@ -204,6 +204,151 @@ def test_scc_base_resolve_vector_notation_config(frontend, horizontal):
...
+    fcode_driver = """
+  SUBROUTINE drive_scalar_rhs(start, end, nlon, q, scalar)
+    INTEGER, INTENT(IN) :: start, end, nlon
+    REAL, INTENT(INOUT) :: q(nlon)
+    REAL, INTENT(IN) :: scalar
+
+    q(start:end) = 0.
+    q(start:end) = scalar
+  END SUBROUTINE drive_scalar_rhs
+"""
```
Comment (mlange05) [inline loki/transformations/single_column/tests/test_scc.py:287]: Is this the same code block as above? Could we use test parametrisation instead here?

**Review by mlange05 — COMMENTED** [review]
The changes in this PR look great and the increased test coverage is very much appreciate. I've flagged a bunch of cosmetics and LLM-isms that would be great to address, as the increased coverage balloons the associated test quite a bit. I'll see if this can be improved before merging, but content-wise it looks good.

**Review by mlange05 — APPROVED** [review]
With the cosmetic changes in this is now GTG for me. A very nice contribution overall - many thanks!
